### PR TITLE
API-4400 add webchat widget support

### PR DESCRIPTION
**Pull request by PrinceNebulon**
Implementing https://developer.mypurecloud.com/api/webchat/widget-version2.html

Fields, code examples, and actual chat initiation are dynamically updated based on if the selected deployment is v1 or v2. V2 deployments will have `(V2)` appended to their name.

### CODE REVIEW

**Comment by PrinceNebulon** [comment]
image

**Comment by idemtidem** [comment]
It'd be best to match the admin UI around the naming conventions, i.e. spelling out "Version 2" or "Version 1" to avoid confusion.

**Comment by PrinceNebulon** [comment]
> It'd be best to match the admin UI around the naming conventions, i.e. spelling out "Version 2" or "Version 1" to avoid confusion.

image

File: app/controllers/webchat.js [hunk 1]
```diff
@@ -33,6 +34,21 @@ export default Ember.Controller.extend({
...
+	webChatVersion: computed('deployment', 'webChatService.deploymentCount', function() {
+		const deploymentId = this.get('deployment');
+		const deployment = this.get('deployments').find((d) => d.id === deploymentId);
+		return !deployment || !deployment.isV2 ? '1' : '2';
+	}),
+	isV2: computed('deployment', 'webChatService.deploymentCount', function() {
+		return this.get('webChatVersion') === '2';
+	}),
+	isV1: computed('deployment', 'webChatService.deploymentCount', function() {
+		return this.get('webChatVersion') !== '2';
```
Comment (asdf123101) [inline app/controllers/webchat.js]: `=== 1`? In case we have v3 in the future.

File: app/controllers/webchat.js [hunk 2]
```diff
@@ -33,6 +34,21 @@ export default Ember.Controller.extend({
...
 	customAttributes: [],
 	error: '',
+	customPlugin: undefined,
+	environmentTld: computed('deployment', function() {
+		return purecloudEnvironmentTld();
+	}),
+	webChatVersion: computed('deployment', 'webChatService.deploymentCount', function() {
+		const deploymentId = this.get('deployment');
+		const deployment = this.get('deployments').find((d) => d.id === deploymentId);
+		return !deployment || !deployment.isV2 ? '1' : '2';
```
Comment (asdf123101) [inline app/controllers/webchat.js]: This seems to be incorrect. If deployment is `undefined` from the line above then this line returns `true`. Do you mean `deployment || !deployment.isV2 ? '1' : '2'`?

File: app/controllers/webchat.js [hunk 3]
```diff
@@ -198,82 +214,183 @@ export default Ember.Controller.extend({
...
 			if(storage.localStorageGet('relate.ui.useEmailAndPhoneForRWPLookupInWebChat')) {
 				chatData.email = this.get('email');
 			}
 
-			const customAttributes = this.get('customAttributes');
-			for (let attribute of customAttributes) {
-				if (attribute.name != '') {
-					chatData[attribute.name] = attribute.value;
+			// Add custom fields if V1
+			if (this.get('isV1') === true) {
```
Comment (asdf123101) [inline app/controllers/webchat.js]: `if (this.get('isV1')`. We know its boolean.

File: app/controllers/webchat.js (hunk 2 above)
Comment (PrinceNebulon) [reply app/controllers/webchat.js]: `!deployment` is intentional. If no deployment is selected, it defaults to displaying v1 data. That also prevents an NPE from the second part of the condition if deployment is undefined.

File: app/controllers/webchat.js (hunk 3 above)
Comment (PrinceNebulon) [reply app/controllers/webchat.js]: Thanks, I think that was leftover after I refactored from doing something else.

**Pull request merged**